PAL_VirtualUnwindOutOfProc for MacOS

mikem8361 opened the pull request:
Add MacOS compact unwind section parsing (__unwind_info) and stepping: SearchCompactEncodingSection, SearchDwarfSection, GetProcInfo, StepWithCompactEncoding*.

Add back some (from a previous commit) of the dwarf unwind info parsing functions to use for the dwarf unwind info (__eh_frame section): ReadValue*, ReadULEB128, ReadSLEB128, ReadEncodedPointer, ParseCie, ExtractFde, ExtractProcInfoFromFde. These functions were used and fairly well tested in 3.x.

Build libunwind8 source on MacOS for those dwarf unwind info cases.

Commit:
PAL_VirtualUnwindOutOfProc for MacOS

Add MacOS compact unwind section parsing (__unwind_info) and stepping: SearchCompactEncodingSection, SearchDwarfSection, GetProcInfo, StepWithCompactEncoding*.

Add back some (from a previous commit) of the dwarf unwind info parsing functions to use for the dwarf unwind info (__eh_frame section): ReadValue*, ReadULEB128, ReadSLEB128, ReadEncodedPointer, ParseCie, ExtractFde, ExtractProcInfoFromFde. These functions were used and fairly well tested in 3.x.

Build libunwind8 source on MacOS for those dwarf unwind info cases.

mikem8361 commented:
@janvorli, @sdmaclea as far as the changes I made to build libunwind8 for MacOS, I hope we can come up with a way to add them without requiring them to be pushed to the libunwind8 repo.  I didn't make any changes to the core files only changed CMakeLists.txt and adding some subdirectories (which could be moved out of the libunwind directory).

Commit:
Move macOS libunwind source changes to libunwind_mac

sdmaclea commented on src/coreclr/src/debug/daccess/dacfn.cpp:
```diff
@@ -277,9 +277,12 @@ DacVirtualUnwind(ULONG32 threadId, PT_CONTEXT context, PT_KNONVOLATILE_CONTEXT_P
-    else
 #endif
+
+    if (hr == E_NOINTERFACE || hr == E_NOTIMPL)
```
This looks fishy.  What if `!FEATURE_DATATARGET4` and `SUCCEEDED(hr)`

mikem8361 replied:
Then hr is E_NOINTERFACE and PAL_VirtualUnwindOutOfProc is called.

What SUCCEEDED(hr) are you looking at? 

```

    HRESULT hr = E_NOINTERFACE;           

#ifdef FEATURE_DATATARGET4
    ReleaseHolder<ICorDebugDataTarget4> dt;
    hr = g_dacImpl->m_pTarget->QueryInterface(IID_ICorDebugDataTarget4, (void **)&dt);
    if (SUCCEEDED(hr))
    {
        hr = dt->VirtualUnwind(threadId, sizeof(CONTEXT), (BYTE*)context);
    }
#endif

    if (hr == E_NOINTERFACE || hr == E_NOTIMPL)
    {
        hr = S_OK;

        SIZE_T baseAddress = DacGlobalBase();
        if (baseAddress == 0 || !PAL_VirtualUnwindOutOfProc(context, contextPointers, baseAddress, DacReadAllAdapter))
        {
            hr = E_FAIL;
        }
    }
```

sdmaclea commented on src/coreclr/src/pal/src/exception/compact_unwind_encoding.h:
```diff
@@ -0,0 +1,478 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
Is this correct?  Did we edit LLVM's file?  Do we need to add the .NET foundation header?

mikem8361 replied:
I added the MIT license.  Not sure if I should have or not.

mikem8361 replied:
I'll remove the MIT header I added.

sdmaclea commented on src/coreclr/src/pal/src/libunwind/src/CMakeLists.txt:
```diff
@@ -332,11 +343,10 @@ else(CLR_CMAKE_HOST_UNIX)
     add_library(libunwind_xdac
       OBJECT
       ../../exception/remote-unwind.cpp
-      ../../exception/seh-unwind.cpp
```
I assume this is intentional and the cross DAC still compiles.

mikem8361 replied:
It does.  The reason is I added a copy of the two functions remote-unwind.cpp needed from seh-unwind.cpp to remote-unwind.cpp.

sdmaclea commented on src/coreclr/src/pal/src/libunwind_mac/include/elf.h:
```diff
@@ -0,0 +1,10 @@
+// This is an incomplete & imprecice implementation
+// It defers to the open source freebsd-elf implementations.
```
Some of these files appear to be direct clones of the Windows platform abstraction files I added to make Windows libunwind compile.  I'm not sue how I feel about cloning them again.

I guess it really depends on whether we plan to upstream this.

mikem8361 replied:
The windows versions of these duplicate files are ifdef'ed _MSC_VER and I didn't want to change them.  We can fix that if we do actually ever upstream these changes.

sdmaclea commented on src/coreclr/src/pal/src/libunwind_mac/src/missing-functions.c:
```diff
@@ -0,0 +1,42 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
I wonder if this should use libunwind's license header.

mikem8361 replied:
I'm not sure either.  I did copy some of those function from the libunwind source.

mikem8361 replied:
I removed the MIT header and added the libunwind one.

sdmaclea reviewed (commented):
It is not inherently obvious why we wouldn't want to upstream this.  Perhaps because MacOS libunwind support is always installed.  So there is no need for MacOS support in the libunwind project.

I have mixed feelings about the directory layout.  I would wait for @janvorli feedback on the directory style.

mikem8361 commented:
As far as push these changes to the libunwind repo: it would require a lot of work because most of the time the dwarf unwind isn't used so libunwind isn't used.  It is the compact/compressed unwind info that is used at lot more and requires the compact "step" code I wrote in remote-unwind.cpp. So most of the new code would have to be ported to the libunwind8 coding conventions, etc.  In other words, without the compact format parsing/stepping just building it on macOS  doesn't have any benefit.

janvorli commented on src/coreclr/src/pal/src/libunwind_mac/include/endian.h:
```diff
@@ -0,0 +1,9 @@
+// This is an incomplete & imprecice implementation of the
```
Do we need these fake files for Mac at all? On Windows, we didn't have some of the headers / symbols / functions. But on OSX, I would expect most if not all of those being present in the platform headers.

mikem8361 replied:
Some of them like elf.h doesn't exist on macOS. Others, in order to build libunwind without any changes are needed to fake it out.

mikem8361 commented on src/coreclr/src/pal/src/exception/remote-unwind.cpp:
```diff
@@ -152,6 +160,1254 @@ typedef struct _libunwindInfo
+
+#define EXTRACT_BITS(value, mask)       ((value >> __builtin_ctz(mask)) & (((1 << __builtin_popcount(mask))) - 1))
+
+// DWARF Pointer-Encoding (PEs).
```
I added a comment.

mikem8361 commented on src/coreclr/src/pal/src/CMakeLists.txt:
```diff
@@ -226,6 +233,22 @@ add_library(coreclrpal
+
+  set_source_files_properties(${CLR_DIR}/src/pal/src/exception/remote-unwind.cpp PROPERTIES
+    COMPILE_FLAGS "-D_XOPEN_SOURCE -DUNW_REMOTE_ONLY"
+    INCLUDE_DIRECTORIES "${CLR_DIR}/src/pal/src/libunwind/include;${CLR_DIR}/src/pal/src/libunwind/include/tdep;${CMAKE_CURRENT_BINARY_DIR};${CMAKE_CURRENT_BINARY_DIR}/libunwind/include;${CMAKE_CURRENT_BINARY_DIR}/libunwind/include/tdep"
```
I'll make these changes.

Commit:
Code review feedback

janvorli commented on src/coreclr/src/pal/src/CMakeLists.txt:
```diff
@@ -226,6 +233,26 @@ add_library(coreclrpal
+  )
+
+  target_include_directories(coreclrpal_dac PUBLIC 
+    "${CLR_DIR}/src/pal/src/libunwind/include;${CLR_DIR}/src/pal/src/libunwind/include/tdep;${CMAKE_CURRENT_BINARY_DIR};${CMAKE_CURRENT_BINARY_DIR}/libunwind/include;${CMAKE_CURRENT_BINARY_DIR}/libunwind/include/tdep"
```
Just a final nit - can you please put one directory per line here instead of the semicolon separated list and use `CMAKE_CURRENT_SOURCE_DIR` instead of `${CLR_DIR}/src/pal/src` as follows?
```suggestion
    ${CMAKE_CURRENT_SOURCE_DIR}/libunwind/include
    ${CMAKE_CURRENT_SOURCE_DIR}/libunwind/include/tdep
    ${CMAKE_CURRENT_BINARY_DIR}
    ${CMAKE_CURRENT_BINARY_DIR}/libunwind/include
    ${CMAKE_CURRENT_BINARY_DIR}/libunwind/include/tdep```

mikem8361 replied:
Done.

janvorli reviewed (approved):
LGTM modulo the nit.

Commit:
More code review feedback and fix "clrstack -i" failure

The methoddesc's enum memory region code needed add more of the code version manager's data to the coredump.
